### PR TITLE
bn: Properly error out if aliasing return value with modulus

### LINKED ISSUE #21110: BN_mod_inverse incorrect result when parameters are aliased
```c
#include <openssl/bn.h>

#define CF_CHECK_NE(expr, res) if ( (expr) == (res) ) { goto end; }

int main(void)
{
    char* str = NULL;
    BIGNUM* a = BN_new();
    BIGNUM* b = BN_new();
    BN_CTX* ctx = BN_CTX_new();
    CF_CHECK_NE(BN_dec2bn(&a, "5193817943"), 0);
    CF_CHECK_NE(BN_dec2bn(&b, "3259122431"), 0);
    CF_CHECK_NE(BN_mod_inverse(b, a, b, ctx), NULL);
    str = BN_bn2dec(b);
    printf("%s\n", str);
end:
    BN_free(a);
    BN_free(b);
    BN_CTX_free(ctx);
    OPENSSL_free(str);
    return 0;
}
```

This prints 0 but should print 2609653924. Note that in the call to `BN_mod_inverse`, the result is the same pointer as the modulus. The documentation explicitly states that "r may be the same BIGNUM as a or n.".

This also affects @davidben and @botovq.

### PR DESCRIPTION
**Pull request by t8m**
Fixes #21110

This is alternative to #21124 as requested by OTC.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/bntest.c
Comment (botovq) [inline test/bntest.c:3231]: Probably worth adding a check with `BN_mod_exp_simple(b, a, b, c, ctx)` since you fix aliasing of `r` with `d`.

Other than that this PR looks good to me.

**Comment by botovq** [comment]
`BN_mod_exp_recp()` needs a similar fix as `BN_mod_exp_simple()` for aliasing r with p. The (3, 1, 11) testcase is too simple to catch that since it happens to return 3 in that situation. You could use (3, 2, 11) with a result of 9 instead, or better yet use a larger exponent where the calculation actually needs to take a non-trivial modulus.

Aliasing `r` with `m` happens to work because of the way `m` is used in `BN_mod_exp_recp()`.

**Comment by t8m** [comment]
> `BN_mod_exp_recp()` needs a similar fix as `BN_mod_exp_simple()` for aliasing r with p. The (3, 1, 11) testcase is too simple to catch that since it happens to return 3 in that situation. You could use (3, 2, 11) with a result of 9 instead, or better yet use a larger exponent where the calculation actually needs to take a non-trivial modulus.
> 
> Aliasing `r` with `m` happens to work because of the way `m` is used in `BN_mod_exp_recp()`.

Fixup pushed

File: test/bntest.c
Comment (t8m) [reply test/bntest.c:3239]: Done

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
ping @openssl/committers for second review

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by hlandau** [comment]
Adding 3.2 on the basis that any approved PR targeting master and at least one other branch should be assumed a candidate for 3.2.

**Comment by hlandau** [comment]
Merged to 3.0, 3.1, 3.2 and master. Thank you.

**Pull request closed without merging**
